azkaban-common/src/main/java/azkaban/project/AzkabanProjectLoader.java

```diff
@@ -80,8 +80,46 @@
       throws ProjectManagerException {
     log.info("Uploading files to " + project.getName());
 
-    // Unzip.
-    File file = null;
+    // Since props is an instance variable of ProjectManager, and each
+    // invocation to the uploadProject manager needs to pass a different
+    // value for the PROJECT_ARCHIVE_FILE_PATH key, it is necessary to
+    // create a new instance of Props to make sure these different values
+    // are isolated from each other.
+    final Props prop = new Props(this.props);
+    prop.putAll(additionalProps);
+
+    // Unzip the project.
+    final File file = unzipProject(archive, fileType);
+
+    // Validate the project, load external validators to do the validation.
+    final Map<String, ValidationReport> reports =
+        validateProject(project, archive, file, prop);
+
+    // Todo jamiesjc: in Flow 2.0, we need to create new flowLoader class and
+    // call new method to load the project flows.
+    // Need to guicify it later so that we can mock flowLoader in the tests.
+    // Load the project flows.
+    final DirectoryFlowLoader directoryFlowLoader = new DirectoryFlowLoader(prop);
+    reports.put(DIRECTORY_FLOW_REPORT_KEY,
+        directoryFlowLoader.loadProject(project, file));
+
+    // Check the validation report.
+    if(!isReportStatusValid(reports, project, file)) {
+      return reports;
+    }
+
+    // Upload the project to DB and storage.
+    persistProject(project, archive, uploader);
+
+    // Clean up project directories and old installations.
+    cleanUpProject(project, file);
```

Comment:
should cleanup be in finally block?

Reply:
I can put deleting temp directory to a finally block. But for cleaning up old project installations, it should happen only when uploading new project is successful.